doc: Improve dependencies documentation

NicolaLS opened the pull request:
Initially there was a distinction between the compiler dependencies and
other required dependencies (refs https://github.com/bitcoin/bitcoin/pull/23565) but the distinction was
removed (refs https://github.com/bitcoin/bitcoin/pull/24585) which is why having two distinct tables could lead
to confusion now.

NicolaLS committed:
doc: Improve dependencies documentation

- Move "Clang" and "GCC" from the table to a new "Compiler" heading,
  indicating either is required.
- Move "CMake" into the required table.
- Move "Python" into the optional table.
- Merge the optional dependencies into one table. Removed sub-headers
  are put into parentheses behind the dependency name in the first
  column.
- Replace the whitespace in the "Minimum required" column of "qrencode"
  with "N/A" for consistency.
- Add missing info for the "systemtap" dependency.
- Add context for "Linux Kernel" dependency in parentheses behind the
  dependency name.

hodlinator commented:
Would be good to move discussion/question-type commentary out of the PR summary into its own comment as it becomes part of the commit message, regarding the "I was a bit confused ... please let me know ...". Alternatively re-word it.

fanquake commented:
Not sure. Re-arranging this to put two different compilers under "required" doesn't seem less confusing. If anything, I think the tables should be left as-is, and a "developer tools" header added, if there is any confusion.

NicolaLS commented:
> Not sure. Re-arranging this to put two different compilers under "required" doesn't seem less confusing. If anything, I think the tables should be left as-is, and a "developer tools" header added, if there is any confusion.

With a "development tools" header it would still be unclear which development tools are required (`cmake`, `python`), optional (`systemtap`) or mutual substitutes (`gcc` _or_ `clang`) and since required ones would mixed with the two different compilers in the same table it would have the same issue that you pointed out imo.

Would it make sense to add a "Compiler" header with a paragraph that one of them is required and a table containing both. And then merge the other tables as @maflcko suggested ?

NicolaLS commented:
I agree that the previous version or this PR did not make sense for clearing up the confusion. I force pushed a new approach and amended the commit message / PR description.

hodlinator reviewed (approved):
ACK d0ca5c53635916868e7d5da913faf3d9a8d706b7

Good to clarify that only one of the compilers is required, and re-introduce that section!

Thanks for updating the PR summary.

### Commit message

Suggest removing repetition of initial sentence and fixing grammar of last line, + less important other changes.

```diff
-doc: improve dependencies documentation
+doc: Improve dependencies documentation
 
-This commit improves the dependencies documentation by:
-- moving "Clang" and "GCC" from the table to a new "Compiler" heading.
-- moving "Python" and "Cmake" into the required table.
-- merging the optional dependencies into one table. Removed sub-headers
+- Move "Clang" and "GCC" from the table to a new "Compiler" heading, indicating either is required.
+- Move "Python" and "CMake" into the required table.
+- Merge the optional dependencies into one table. Removed sub-headers
   are put into parentheses behind the dependency name in the first
   column.
-- replacing the whitespace in the "Minimum required" column of
+- Replace the whitespace in the "Minimum required" column of
   "qrencode" with "N/A" for consistency.
-- adding missing info in for the "systemtap" dependency.
+- Add missing info for the "systemtap" dependency.
```

hodlinator reviewed (approved):
re-ACK 1fcc1ce0edcc8ae522ab9103ad79adf58fc0fe0e

Thanks for applying my suggestions.

NicolaLS commented on doc/dependencies.md:
```diff
@@ -4,40 +4,34 @@ These are the dependencies used by Bitcoin Core.
 | Dependency | Releases | Version used | Minimum required | Runtime |
 | --- | --- | --- | --- | --- |
+| CMake | [link](https://cmake.org/) | N/A | [3.22](https://github.com/bitcoin/bitcoin/pull/30454) | No |
+| Python (scripts, tests) | [link](https://www.python.org) | N/A | [3.10](https://github.com/bitcoin/bitcoin/pull/30527) | No |
```
ah yes..thanks. should I move [Linux Kernel](https://github.com/bitcoin/bitcoin/blob/master/doc/dependencies.md?plain=1#L22) to optional too then ?

rkrux commented on doc/dependencies.md:
```diff
@@ -4,40 +4,34 @@ These are the dependencies used by Bitcoin Core.
+## Compiler
+
+Bitcoin Core requires one of the following compilers.
 | Dependency | Minimum required |
 | --- | --- |
 | [Clang](https://clang.llvm.org) | [16.0](https://github.com/bitcoin/bitcoin/pull/30263) |
-| [CMake](https://cmake.org/) | [3.22](https://github.com/bitcoin/bitcoin/pull/30454) |
 | [GCC](https://gcc.gnu.org) | [11.1](https://github.com/bitcoin/bitcoin/pull/29091) |
-| [Python](https://www.python.org) (scripts, tests) | [3.10](https://github.com/bitcoin/bitcoin/pull/30527) |
-| [systemtap](https://sourceware.org/systemtap/) ([tracing](tracing.md))| N/A |
```
This doesn't show up as a table when I preview the md file.

rkrux replied:
Needs a blank line after line 9.

NicolaLS replied:
thanks, I was wondering if there should be a blank line. for me it shows the table correctly (firefox, gh mobile) so I thought when in doubt just keep it compact.

weird that it displays correctly for me but not for you though (also tested with `gh-markdown-preview` locally).

NicolaLS commented:
added a blank line between the paragraph and the table in `## Compiler` and moved `Python` to the `## Optional` table. thanks for the help everyone.

I think `Linux Kernel` should be moved to optional as well then but since it was in the required table before and I'm not sure if I should change more stuff after receiving acks I just left it as is now.

NicolaLS commented:
> Maybe a separate table for runtime dependencies makes more sense. It is pretty clear that runtime dependencies don't have a "Version used". An alternative would be to just remove the "Runtime" column, as it can be derived from the "Version used" column.

tbh. I don't understand what "version used" really means in this context, looked at a few linked PR's but they went over my head...

but anyways I think this is **out of scope for this PR** / could be addressed in another PR. I only wanted to clear up some confusion I had because of the two separate tables.

hebasto commented on doc/dependencies.md:
```diff
@@ -4,40 +4,35 @@ These are the dependencies used by Bitcoin Core.
 
 | Dependency | Releases | Version used | Minimum required | Runtime |
 | --- | --- | --- | --- | --- |
+| CMake | [link](https://cmake.org/) | N/A | [3.22](https://github.com/bitcoin/bitcoin/pull/30454) | No |
```
CMake 3.24.2 is used for release binaries in the Guix environment.

hodlinator replied:
The plot thickens - reference to lower version than minimum under *depends/*:
https://github.com/bitcoin/bitcoin/blob/df8bf657450d5383870d40790ea9f4fdb03c360d/depends/hosts/linux.mk#L44-L45

I agree with the author in https://github.com/bitcoin/bitcoin/pull/31634#issuecomment-2589465760 that it's fine to limit the scope of the current PR.

hebasto replied:
If you imply that 3.17.0 is a CMake version, it is not. It’s `CMAKE_SYSTEM_VERSION`.

NicolaLS replied:
thank you. so the PR to reference for the "version used" would be #27172 correct ?

[manifest](https://github.com/bitcoin/bitcoin/blob/master/contrib/guix/manifest.scm) uses `((gnu packages cmake) #:select (cmake-minimal))` which is why we can find the version ([relevant comment in cmake.scm](https://git.savannah.gnu.org/cgit/guix.git/tree/gnu/packages/cmake.scm?id=713ac1fe47de1a5f29f6186e51d9ffb2b8083258#n164))

hebasto replied:
>so the PR to reference for the "version used" would be #27172 correct ?

That refers to a Darwin-specific change. https://github.com/bitcoin/bitcoin/pull/29725 is a PR, which made `cmake-minimal` a global requirement.

However, in the future, updates like https://github.com/bitcoin/bitcoin/pull/30730 might bump the `cmake-minimal` version.

hodlinator reviewed (approved):
re-ACK 05a008c1d07a38f589536661822d5450aa4e9438

- Thanks to @rkrux for [spotting the formatting issue](https://github.com/bitcoin/bitcoin/pull/31634#discussion_r1913559740) I missed.
- SQLite now categorized as `(wallet)`. :+1:

hodlinator reviewed (approved):
re-ACK a759ea3e920dcba52798755ba9f3891f914afbb0

`git range-diff master 05a008c a759ea3`

Resolved conflict with recently merged Qt depends upgrade, #30774.

rkrux reviewed (approved):
ACK a759ea3e920dcba52798755ba9f3891f914afbb0

achow101 commented:
ACK a759ea3e920dcba52798755ba9f3891f914afbb0

fanquake commented on doc/dependencies.md:
```diff
@@ -4,40 +4,35 @@ These are the dependencies used by Bitcoin Core.
 | [libevent](../depends/packages/libevent.mk) | [link](https://github.com/libevent/libevent/releases) | [2.1.12-stable](https://github.com/bitcoin/bitcoin/pull/21991) | [2.1.8](https://github.com/bitcoin/bitcoin/pull/24681) | No |
 | glibc | [link](https://www.gnu.org/software/libc/) | N/A | [2.31](https://github.com/bitcoin/bitcoin/pull/29987) | Yes |
-| Linux Kernel | [link](https://www.kernel.org/) | N/A | [3.17.0](https://github.com/bitcoin/bitcoin/pull/27699) | Yes |
+| Linux Kernel (if building that platform) | [link](https://www.kernel.org/) | N/A | [3.17.0](https://github.com/bitcoin/bitcoin/pull/27699) | Yes |
```
What does this mean? I'm not sure it's correct.

hodlinator replied:
Suggested by me here: https://github.com/bitcoin/bitcoin/pull/31634#discussion_r1912232236
Maybe stating the obvious a bit too much for your taste, or how would it be incorrect?

hodlinator replied:
> What does this mean?

My point is that it's not required on other operating systems.

fanquake replied:
> My point is that it's not required on other operating systems.

Right, but it's also not required if cross-compiling to Linux from other operating systems, and yea, if you're self-compiling on Linux it doesn't really matter. This whole line could probably just be dropped.

NicolaLS replied:
I'll go through this PR and check what I can add in a follow up PR. Regarding https://github.com/bitcoin/bitcoin/pull/31634#issuecomment-2589412550 and the discussion on Linux Kernel I came to the following conclusions:
- it does not make sense to mix runtime deps. that are either built from [depends](https://github.com/bitcoin/bitcoin/blob/master/depends/README.md) or come from the system in the same table as required build deps.
- there should be a separate section for runtime deps. that **does not** list the versions used for each package in depends but rather explains what depends does, how to use it, but that it is not required as system packages can be used.
- drop `Runtime` column as there will be a section for runtime dependencies which means all other deps. are "no".
- drop `Version Used` for runtime deps. but document how depends works and that each package defines a version for the deterministic builds. it does not make sense to document every package individually, users interested in the version can easily check if they understand how depends works.
- drop `Linux Kernel` users should not assume they can just build on a EOL Kernel

I'll open a draft PR and link it here.

NicolaLS replied:
Here's the draft: https://github.com/bitcoin/bitcoin/pull/31895

Please let me know if this addresses the Linux Kernel problem and other stuff that was out of scope for this PR sufficiently and if its ready to be reviewed @hodlinator

hodlinator replied:
Looks pretty ready to me, PR description and commits are informative. Left an additional suggestion.
